Wrap binaries under install/lib/{PACKAGE}, not top level

jprestwo opened the pull request:
The current code was wrapping every binary located under install/bin, which contains binaries from packages that may not be related to ament_virtualenv at all. This was causing failures due to the venv being used when it shouldn't.

In addition the use_system_packages was hard coded to false which requires every single python package be within the venv, which is not generally the case.

jprestwo committed:
Utilize distutils to simplify/fix the script wrapping

The current code was wrapping every binary located under
install/bin, which contains binaries from packages that may not
be related to ament_virtualenv at all. This was causing failures
due to the venv being used when it shouldn't.

Instead of having each package define its own install class and
call venv_install, create an AmentVirtualenvInstall class that
can be imported and used by packages. All the information we need
is stored by distutils so there is no need for the package to
define it. This install class will also parse the entry_points and
only wrap scripts defined there, thereby isolating the venv only
to scripts installed by the package.

Extra options are now possible, similar to catkin_virtualenv by
defining an "ament_virtualenv" dictionary within the options
argument to setup().

setup(
    # ...
    options={
        "ament_virtualenv": {
            "use_system_packages": True,
            "python_version": "3.10"
        }
    }
}

paulbovbel commented on ament_virtualenv/ament_virtualenv/install.py:
```diff
@@ -148,7 +148,7 @@ def install_venv(install_base, package_name, python_version='3'):
             python_version=python_version,
             requirements_filename=generated_requirements,
-            use_system_packages=False,
+            use_system_packages=True,
```
in catkin_virtualenv, this is driven via cmake https://github.com/locusrobotics/catkin_virtualenv?tab=readme-ov-file#additional-cmake-options

jprestwo replied:
We could do something similar... except my new world order doesn't lend itself nicely to passing arguments with pure python packages. I made everything "automatic" by defining the install class in ament_virtualenv...

jprestwo replied:
@paulbovbel I figured out how we can supply options within setup.py like we do with catkin_virtualenv. This is taken from the new commit description:

```
Extra options are now possible, similar to catkin_virtualenv by
defining an "ament_virtualenv" dictionary within the options
argument to setup().

setup(
    # ...
    options={
        "ament_virtualenv": {
            "use_system_packages": True,
            "python_version": "3.10"
        }
    }
}
```
For the cmake case I think it would be no different than catkin_virtualenv, just provide the options to the cmake function.

jprestwo commented:
For context this would be the changes for packages depending on ament_virtualenv:
locus_py: https://github.com/locusrobotics/locus_common/pull/218
aiorclpy: https://github.com/locusrobotics/aiorclpy/pull/4

jprestwo committed:
Add typing for arguments/return values

This should be done going forward as changes are made to improve
code readability.

bearrito reviewed (approved):
Looks good. Talked offline. Limiting venv creation to console scripts make sense to me. 
I didn't review the rest of the codebase.

Thanks for adding the typing. That's helpful.
